### PR TITLE
Repro for #378

### PR DESCRIPTION
**Pull request by ca-ta**
Repro for #378: Run ```SMSDeactivationTest```

Notes:

* The test is marked as BVT, but it takes 3+ minutes to execute - you may want to relegate it to nightly runs only
* I haven't created a similar test for the AQ streams - it's trivial to add one, though.
* The test shows that the deactivation cycle hangs and that subsequent calls to one of the grains involved in the deadlock hang. Adding more tests assessing the extent of the silo degradation may make sense (e.g. trying to activate an unrelated grain, etc.).
* I've observed that if one waits further (minutes?) the silo eventually commits suicide.

### CODE REVIEW

**Comment by gabikliot** [comment]
Ok, this is a cool bug!
**Your repro was definitely VERY useful and helped a lot!**

There is a number of different issues here.
There is definitely a problem with 2 activations being collected in the same time in this particular scenario. There is a bug and I think I know how to address it. But lets put that on a side for a while and first address the most worrisome issue: you said that it appears as silo is stuck after the deactivation cycle, is unresponsive and eventually commits suicide.
This is of course very bad and should not happen.

I believe what you are seeing is the consequence of this particular test scenario. **The silo works fine, the test just looks like the silo is dysfunctional.**
After grain B (the producer grain) is collected, you send it a new message to produce again. This grain is rightfully reactivated again (I added printing of OnActivateAsync and saw it being reactivated). Inside the Fire method you get null ref exception, since producer object is null, as expected. The call throws, it throws back on the client. The test is finished (with error in this case, but it does not matter) and the silo is just stopping, as usual, after the test is done.

So first thing lets make sure the silo itself is OK. If after the long timeout you send msgs to other grains, or send a msg to the same producer grain asking it to become producer again, everything should work fine. The grain will recover and will be reactivated. **Eventually**, immediately after this one deactivation timeout, everything is back to normal. In fact, in this particular case, the deactivation deadlock is not seen by the application. There is no visible application error in this particular case. All works OK as excepted (there is still an internal bug that will surface in other case and we will fix it).

Unrelated, there was also another issue, which I just fixed in #383. The way you set the GC setting caused memory store grain to be collected as well. It does not impact the silo, but pub sub grain gets an empty state after reactivated, which also potentially complicated the test logic as well, since the consumer is now lost.

So please try with the updated test and #383 and lets validate silo is fine. After that we will discuss fixing the deactivation deadlock issue.

**Comment by ca-ta** [comment]
@gabikliot - thank you for looking into it!

About the bug - from my cursory look at the code, a naïve solution seemed to be pulling the grains that "use streams" into a preliminary deactivation  - I didn't want to make changes in that area though, as it is a critical piece of code that I'm not familiar with. 

Sorry about missing the call to ```BecomeProducer``` after the long timeout - it shows I shouldn't write tests after a few Sunday night beers :-)
I will modify the test locally - I do not have access to the changes you made to the test (I've looked into your branch as well, but could not find them) and see what that yields. 

Regarding #383 - indeed. And you are right, I've tested with the Azure Table storage as well and the deadlock still occurs.

I'll give you an update after I modify the test. Thank you!

**Comment by gabikliot** [comment]
Don't worry about the bug itself, for now. I have ideas how to fix it and will for sure post an update about why this happens and how I think we can fix it.
As I said in this particular case of PubSubGrain and Producer, the deactivation deadlock is actually benign. You should not see any exception on the application level, and if the test handles all correctly (BecomeProducer) it should work even without the deadlock fix. Lets make sure about that first.

I just want to make 100% sure, without any doubt, that what ever individual grain calls are timing out, failing, deadlocking, the silo itself is not impacted, it stays healthy and can serve other grains and also the impacted grains recover and can process new messages. This is exactly one of the main points about **Virtual Actor abstraction** - **no matter what happens, we eventually recover from any problem and keep going**, without application needing to write any recovery code. In this particular case there is a nasty deadlock between actors, but we recover, create a new actor activation and keep going as if nothing happened.

I have not done any test changes, merely added some logging in the grains, saw that all works as I expected.

**Comment by ca-ta** [comment]
@gabikliot  - I've done some more digging, and I've just pushed an improved version of the tests. Now we have 2 tests, one for a grain subscriber and another one for a client subscriber (the later closer matches my scenario).

The good news : *with the current silo test settings*, the tests behave as expected - the silo recovers and the tests pass.

Now, comment out ```<Messaging ResponseTimeout="30s"/>``` from ```OrleansConfigurationForStreamingDeactivationUnitTests.xml``` - the tests never finish. Or maybe the silo does eventually restart (as I think I've observed to be the case for may system), but I am writing this without having waited for that to happen.

It appears that that was exactly why my system ground to a halt - in the absence of the ```<Messaging ResponseTimeout="***"/>``` entry from the Silo configuration, any subsequent request tot the producer grain hangs.

In conclusion, I guess the silo becoming unresponsive in terms of calls to hung grains could be attributed to user error (my not setting the ```<Messaging ResponseTimeout="***"/>``` in the silo configuration). That being said, perhaps a more forgiving default silo configuration and/or documentation regarding this setting (as well as the client-side counterparts, including ```ClientSenderBuckets``` and ```MaxResendCount```) cannot hurt.

And yes, It is certainly true that I would have not caught the deadlock bug if the configuration was more forgiving. I'm thinking though that bug-surfacing configuration settings lie with the test harness, rather than with the silo defaults.

Thank you for your comments and observations - they helped a lot!

**Comment by gabikliot** [comment]
Catalin, are you running with debugger attached? With debugger attached default timeout is artificially increased to minutes (don't remember how much but a lot). Without debugger it is 30 sec by default.

**Comment by ca-ta** [comment]
Yes, I was, and I see how it makes sense to help debugging by increasing the response timeout.
I've probably unwittingly benefitted from that through countless debug sessions - up until the deadlock bug, that is :-)

**Comment by gabikliot** [comment]
Yes, exactly. And that is why it appears as if the call is stuck. But if you run without debugger, the call will timeout after 30 sec.

**Comment by gabikliot** [comment]
@ca-ta, so both those tests are passing now, if run under BVT, without debugger attached?
I would like to merge them in our test suite, even before we go into fixing the deactivation deadlock.

**Comment by ca-ta** [comment]
Yes, they indeed pass.

Hold off the merge though, I am making them catch the deadlock bug - so they will soon fail, for the right reasons. I think we should merge them after I make the changes.

**Comment by ca-ta** [comment]
Also, I was thinking of removing the BVT category for now, as they take a couple of minutes each... We could re-introduce it when/if we have a way to trigger deactivation on-demand and reduce the test running time.
What is your take on this?

**Comment by gabikliot** [comment]
Sure thing, won't merge. 
And OK for removing BVT.

**Comment by ca-ta** [comment]
Ready to merge:
* I've set the ```ResponseTimeout``` to 1hr in ``` OrleansConfigurationForStreamingDeactivationUnitTests.xml``` (it only affects these two tests).
  * This should take care of the discrepancies between running the tests with the debugger attached and without.
*  I've modified the tests themselves to detect that the producer grain is hung as a result of the deactivation bug.
  * Both tests now fail for the right reason (comment out the long sleep, and they both pass)

This really is a funky bug - sorry for the silo scare :-)

**Comment by gabikliot** [comment]
With fixing the original bug via https://github.com/dotnet/orleans/pull/399 I also set those 2 tests to expect success, instead of failure as before.
